File: Data Structures & Algorithms/count-number-of-islands/submission-1.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        
        rows = len(grid)
        cols = len(grid[0])
        island = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    island += 1
                    self.dfs(grid, i, j, rows, cols)
        
        return island
    
    def dfs(self, grid, r, c, rows, cols):
        if r < 0 or c < 0 or r >= rows or c >= cols:
            return
        
        if grid[r][c] == "0":
            return 
        
        grid[r][c] = "0"

        direction = [
            (-1,0),
            (1,0),
            (0,-1),
            (0,1)
        ]

        for dr, dc in direction:
            self.dfs(grid, r+dr, c+dc, rows, cols)

File: Data Structures & Algorithms/islands-and-treasure/submission-1.go
func islandsAndTreasure(grid [][]int) {
    rows := len(grid)
	cols := len(grid[0])

	for i:=0;i<rows;i++{
		for j:=0;j<cols;j++{
			if grid[i][j] == math.MaxInt32{
				dist := bfs(grid,i,j,rows,cols)

				if dist != math.MaxInt32{
					grid[i][j] = dist
				}
			}
		}
	}
}

func bfs(grid [][]int, r, c, rows, cols int)int{
	queue := [][]int{{r,c,0}}
	visited := make(map[[2]int]bool)
	visited[[2]int{r,c}] = true

	for len(queue) > 0{
		current := queue[0]
		queue = queue[1:]

		row := current[0]
		col := current[1]
		dist := current[2]

		if grid[row][col] == 0{
			return dist
		}

		directions := [][]int{
			{1,0},
			{-1,0},
			{0,-1},
			{0,1},
		}

		for _, dir := range directions{
			newR := row+dir[0]
			newC := col+dir[1]

			if newR >= 0 && newR < rows && newC >=0 && newC < cols && !visited[[2]int{newR,newC}] && grid[newR][newC] != -1{
				visited[[2]int{newR,newC}] = true
				queue = append(queue, []int{newR,newC,dist+1})
			}
		}
	}

	return math.MaxInt32
}